csharp/extractor/Semmle.Extraction.CSharp/Entities/LocalVariable.cs

```diff
@@ -90,6 +90,6 @@ class LocalVariableFactory : ICachedEntityFactory<(ISymbol, Expression, bool, Ex
                 new LocalVariable(cx, init.Item1, init.Item2, init.Item3, init.Item4);
         }
 
-        public override TrapStackBehaviour TrapStackBehaviour => TrapStackBehaviour.NeedsLabel;
+        public override TrapStackBehaviour TrapStackBehaviour => TrapStackBehaviour.OptionalLabel;
```

Comment:
Can you elaborate on why this is needed? I wonder because the `Field` entity [sets](https://github.com/Semmle/ql/blob/master/csharp/extractor/Semmle.Extraction.CSharp/Entities/Field.cs#L99) `TrapStackBehaviour.PushLabel`, which I assume happens before extracting the initializer?

Reply:
The reason is because fields don't use the tag stack. Most of the time, local variables are declared inside methods, or in lambdas in methods, so the tag stack is non-empty. However, a local variable can be declared inside a lambda which is used as an initializer to a field, e.g. [here](https://github.com/Semmle/ql/blob/1ac5660e2c58a083ddf721409fc242ac084fb930/csharp/ql/test/library-tests/regressions/Program.cs#L133).

In those cases, there is no surrounding method, and therefore the tagstack is empty. This was causing the `tagstack unexpectedly empty` messages in `csharp.log`.

Changing the behaviour to `OptionalLabel` removes that warning.

Reply:
> The reason is because fields don't use the tag stack

Then what does my link from above mean, I must have misunderstood something? (https://github.com/Semmle/ql/blob/master/csharp/extractor/Semmle.Extraction.CSharp/Entities/Field.cs#L99)

Reply:
I reverted this change because it was the wrong fix. It actually turned out that the erroneous case was a *property* with an expression body, so I changed the trap stack behaviour of properties to `PushesLabel`.

Reply:
Excellent, thanks for investigating this!